Add support for setting the heading level for web semantics (#97894)

victorgalo opened the pull request:
This change adds a new property in Semantics widget that would take an integer corresponding to the heading levels defined by the ARIA heading role. This is necessary in order to get proper accessibility and usability in a website for users who rely on screen readers and other assistive technologies.

Issue fixed by this PR:
fixes https://github.com/flutter/flutter/issues/97894

Engine part:
https://github.com/flutter/engine/pull/41435 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

victorgalo committed:
Add support for setting the heading level for web semantics (#97894)

victorgalo committed:
Add support for setting the heading level for web semantics (#97894)

victorgalo commented:
The build is failing because the engine part must be approved. I need to integrate both changes together (framework and engine part) for integration tests to pass. Please, someone can tell me how I should deal with this?.

Maybe @yjbanov can shed some light?

Thank you!

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -535,6 +536,11 @@ class SemanticsData with Diagnosticable {
 
+  /// The hierarchy level of the widget as a header.
+  ///
+  /// This should only be set when [SemanticsFlag.isHeader] is set.
```
is this true? the isHeader may be closer to banner in aria.

victorgalo replied:
Yes, a widget with isHeader attribute gets a "role=heading" aria attribute, not a "banner" role. For that reason I consider this change is needed, a "heading" role needs an "aria-level". Webs without this do not pass accessibility static analysis checks and we really need this for a specific project.

chunhtai replied:
> isHeader attribute gets a "role=heading

is this already the case in web? I remember when I last revisited this, only the Appbar has the isHeader = true. and Appbar is closer to banner in html https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/banner_role

The heading, ie h1, h2, h3 seams to be a different thing in html

cc @yjbanov

chunhtai replied:
This doesn't seem to be answered yet

victorgalo replied:
Hi @chunhtai, sorry for replying so late. Yes, I don't know what's the purpose of having that parameter enabled for the AppBar, but I can confirm applying the "header" attribute has the effect I mention, I tested it many times, also you can see evidence in this code (part of the engine)

![image](https://github.com/flutter/flutter/assets/3387118/5958d4cc-6df3-4d0b-9db1-760d972a9ffe)

chunhtai replied:
I think the code you link is incorrect. @yjbanov should it use banner in engine instead?

yjbanov replied:
Yes, I think flutter web's usage of `role=heading` is incorrect. It was my mistake, as I was misunderstood what `isHeader` meant. "banner" seems closer to what it means in Flutter.

victorgalo replied:
If that's the case and we change isHeader to "banner", there won't be a way to add headings and levels in flutter web. Having a heading structure is an important feature in order to have decent web accessibility. Should we have another attribute for this?

chunhtai replied:
`isHeader` is named after the similar widget in mobile platforms.

https://developer.apple.com/documentation/uikit/uiaccessibilitytraitheader?language=objc

Android called it heading

https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo#isHeading() 

It seems to me the mobile `header` or `heading` is the equivalent `banner` in web.

> Should we have another attribute for this?

Isn't the heading level enough for this? we need to make sure we document that the heading level has nothing to do with the `isHeader` property.

chunhtai replied:
This seems not addressed yet?

victorgalo replied:
> This seems not addressed yet?

Sorry, not yet, but I will address it. If I understood correctly the idea is to separate both attributes, changing "heading" property to be used to establish a "banner" role in web, and dedicate the headingLevel attribute to add the element as a header role with its corresponding level.

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -4437,6 +4471,18 @@ class SemanticsConfiguration {
     _setFlag(SemanticsFlag.isHeader, value);
   }
 
+  /// Indicates the owning [RenderObject]'s hierarchy level.
```
This doc is a bit confusing. what is a hierarchy level?

chunhtai replied:
Can you add more detail to this?

victorgalo replied:
Sure, I tried to add more details. Please let me know if still should add more information. Thank you.

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -2380,7 +2399,17 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
   int? _currentValueLength;
 
-  bool _canPerformAction(SemanticsAction action) => _actions.containsKey(action);
+  /// The level of the widget (when it's defined as a header) within
```
First paragraph should be a brief sentence also this can use more explanation. can this parameter make sense out side of dom or markdown? If not, we should add document in terminologies in dom or markdown

victorgalo replied:
Sure, I tried to add more specific explanation here, please let me know if I should add/change something else. Thank you!

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -1321,6 +1331,14 @@ class SemanticsProperties extends DiagnosticableTree {
   /// [Directionality] or an explicit [textDirection] should be provided.
   final String? tooltip;
 
+  /// The level of the widget (when it's defined as a header) within
```
same

victorgalo replied:
Also changed this, hope it's better now. Thank you.

chunhtai reviewed (commented):
we should also have logic to automatically apply these when people using builtin text style in theme

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

victorgalo committed:
Improve documentation comments for headingLevel semantics property

victorgalo commented:
> we should also have logic to automatically apply these when people using builtin text style in theme

Sorry, I didn't get this. You mean the heading level should automatically be applied with a default value?

chunhtai reviewed (commented):
Hi @victorgalo , are you still planning on coming back to this pr?

chunhtai commented:
> > > we should also have logic to automatically apply these when people using builtin text style in theme
> > 
> > 
> > Sorry @chunhtai, I didn't get this. You mean the heading level should automatically be applied with a default value?
> 
> The Text theme has h1~h6 style (which was replace by displayLarge, displayMedium.. etc). I was wondering where we may have a way to auto apply the semantics headinglevel if these style are in use. One idea would be make TextStyle carry heading level

This can be a separate pr though

victorgalo commented:
> Hi @victorgalo , are you still planning on coming back to this pr?

Yes, sorry for taking so long to reply. I would really appreciate my changes to be included.

chunhtai commented:
Hi @victorgalo are you still planning on working on this pr? The discussion seems to not be resolved yet https://github.com/flutter/flutter/pull/125771#discussion_r1259980025

victorgalo commented:
> Hi @victorgalo are you still planning on working on this pr? The discussion seems to not be resolved yet [#125771 (comment)](https://github.com/flutter/flutter/pull/125771#discussion_r1259980025)

Yes I will, sorry for the delay.

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

# Conflicts:
#	packages/flutter/test/semantics/semantics_update_test.dart

victorgalo committed:
Change documentation adapted to new behaviour for headingLevel

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

victorgalo committed:
Comment integration test temporarily until engine change is done (issue 41435)

victorgalo committed:
Revert changes temporarily until engine part is landed

victorgalo committed:
Comment integration test temporarily until engine change is done (issue 41435)

victorgalo committed:
Comment integration test temporarily until engine change is done (issue 41435)

victorgalo committed:
Comment integration test temporarily until engine change is done (issue 41435)

victorgalo committed:
Comment integration test temporarily until engine change is done (issue 41435)

victorgalo committed:
Comment integration test temporarily until engine change is done (issue 41435)

victorgalo committed:
Comment integration test temporarily until engine change is done (issue 41435)

victorgalo commented:
Hi @chunhtai @yjbanov, I have the engine change ready, but to be able to land it I need first to temporarily "skip" one integration test from the framework part. 

As a first step I pretend to use this PR to just skip the test, when the engine is landed (step 2) I will add the rest of the framework part again and will "unskip" this test(step 3). What I'm not sure is if I could still be using this PR to add the 3rd step changes or I will need to create an additional PR for that.

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

yjbanov commented:
I left some comments on the engine PR. I'm a little confusing about this PR. The sum total of changes now shows that it just skips tests. Should that be done in a separate PR?

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

victorgalo committed:
Add support for setting the heading level for web semantics (#97894)

victorgalo commented:
> I left some comments on the engine PR. I'm a little confusing about this PR. The sum total of changes now shows that it just skips tests. Should that be done in a separate PR?

@yjbanov sorry, my bad, I made this very confusing. I brought back the original changes to this PR and created an additional one to temporarily skip the integration test:
https://github.com/flutter/flutter/pull/135077

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -534,6 +535,12 @@ class SemanticsData with Diagnosticable {
+  ///
+  /// The value should be a number between 1 and 6, indicating the hierarchical
+  /// level as a heading.
+  final int? headingLevel;
```
same here, we should not use null in this class, as there isn't a way to represent null in engine

victorgalo replied:
Done, I'm initially making the attribute non required because otherwise there will be a lot of failures in integration tests. We could make it required later, I hope that works for you :)

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -4528,6 +4563,20 @@ class SemanticsConfiguration {
+  int? get headingLevel => _headingLevel;
+  int? _headingLevel;
+
+  set headingLevel(int? value) {
```
should probably guard around the input range?

victorgalo replied:
Good point, done.

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -436,6 +436,7 @@ class SemanticsData with Diagnosticable {
     required this.platformViewId,
     required this.maxValueLength,
     required this.currentValueLength,
+    required this.headingLevel,
```
add an assert for the range?

victorgalo replied:
Also done.

victorgalo committed:
Change headingLevel to be non nullable
Add guard and assert for headingLevel input values

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

victorgalo committed:
Change headingLevel to be non nullable

victorgalo committed:
Change headingLevel to be non nullable

victorgalo committed:
Change headingLevel to be non nullable

victorgalo committed:
Change headingLevel to be non nullable

victorgalo committed:
Change headingLevel to be non nullable

victorgalo committed:
Change headingLevel to be non nullable

victorgalo committed:
Change headingLevel to be non nullable

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

victorgalo committed:
Remove headingLevel from semantics_udpate_test

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -742,7 +751,8 @@ class SemanticsData with Diagnosticable {
         && other.thickness == thickness
-        && _sortedListsEqual(other.customSemanticsActionIds, customSemanticsActionIds);
+        && _sortedListsEqual(other.customSemanticsActionIds, customSemanticsActionIds)
+        && other.headingLevel == headingLevel;
```
maybe move this above the customaction for code style in this file0

victorgalo replied:
Done.

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -772,6 +782,7 @@ class SemanticsData with Diagnosticable {
       elevation,
       thickness,
       customSemanticsActionIds == null ? null : Object.hashAll(customSemanticsActionIds!),
+      headingLevel,
```
same

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

victorgalo committed:
Change default value for headingLevel to 0

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

victorgalo commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -453,7 +454,8 @@ class SemanticsData with Diagnosticable {
        assert(attributedIncreasedValue.string == '' || textDirection != null, 'A SemanticsData object with increasedValue "${attributedIncreasedValue.string}" had a null textDirection.'),
-       assert(attributedHint.string == '' || textDirection != null, 'A SemanticsData object with hint "${attributedHint.string}" had a null textDirection.');
+       assert(attributedHint.string == '' || textDirection != null, 'A SemanticsData object with hint "${attributedHint.string}" had a null textDirection.'),
+       assert(headingLevel == -1 || (headingLevel >= 1 && headingLevel <= 6), 'Heading level must be between 1 and 6');
```
Ok, changing the default to 0

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -547,6 +549,12 @@ class SemanticsData with Diagnosticable {
 
+  /// Indicates that this subtree represents a heading.
+  ///
+  /// The value should be a number between 1 and 6, indicating the hierarchical
```
maybe mention 0 means this is not a heading

victorgalo replied:
Ok, done.

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -1361,6 +1373,17 @@ class SemanticsProperties extends DiagnosticableTree {
+  /// with aria-level="1", usually indicates the main heading of a page,
+  /// a level 2 heading, defined with aria-level="2" the first subsection,
+  /// a level 3 is a subsection of that, and so on.
+  final int? headingLevel;
```
If this is not null it must be >0 and <=6 right? maybe we should assert it in the constructor

victorgalo replied:
Ok, I added it.

chunhtai reviewed (approved):
LGTM, just nits

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

goderbauer commented:
@victorgalo Do you have plans to address the nits given above in the near future so we can merge this? Also, can you rebase this to the latest main to pick up the engine-side changes, which should make the tests green? Thank you!

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

victorgalo committed:
Add comments and assertions for headingLevel property

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

victorgalo commented:
> @victorgalo Do you have plans to address the nits given above in the near future so we can merge this? Also, can you rebase this to the latest main to pick up the engine-side changes, which should make the tests green? Thank you!
Hi @goderbauer , I addressed the nits, and also merged the latest changes in both framework and engine, but tests are still failing because of infra timeout. I'm willing to finish with these changes but I may need some help here.

victorgalo committed:
Merge branch 'master' into add-semantics-heading-level

chunhtai commented:
@victorgalo can you rebase and try again?

chunhtai commented:
oh the engine needs to be merged first.

chunhtai committed:
Merge branch 'master' into add-semantics-heading-level

chunhtai committed:
fix ci

chunhtai committed:
Merge branch 'master' into add-semantics-heading-level

justinmc reviewed (approved):
LGTM :+1:

k0psutin commented:
I don't think the headingLevel is working?

Try this with MacOS Voice Over: 

https://dartpad.dev/?id=a8e32aa25ea01fe79a590b4a9138687e

Try to select heading with voice over, it says that the Flutter Demo Home Page is the last heading. You can't navigate to the second heading with keyboard arrows either.

chunhtai commented:
Hi @k0psutin it hasn't hook up to desktop embedding i believe. can you open a new issue for this?

k0psutin commented:
Hi @chunhtai I made a new issue #156203
